### PR TITLE
Tooltips

### LINKED ISSUE #367: Tooltips
### Discussed in https://github.com/terrastruct/d2/discussions/329

<div type='discussions-op-text'>

<sup>Originally posted by **azban** December  2, 2022</sup>
It would be nice to have tooltips for objects, so that by default the diagrams are concise, but more information can be encoded for someone looking to dig into specific elements.</div>

### PR DESCRIPTION
**Pull request by alixander**
On SVG, tooltips will be hover to show.

Screen Shot 2022-12-28 at 12 29 54 PM

Exported to PNG, the tooltips will turn into a numbered legend.

Screen Shot 2022-12-28 at 12 25 09 PM

closes #367

### CODE REVIEW

File: d2renderers/d2svg/d2svg.go [hunk 1]
```diff
@@ -39,10 +39,15 @@ const (
 	DEFAULT_PADDING            = 100
 	MIN_ARROWHEAD_STROKE_WIDTH = 2
 	threeDeeOffset             = 15
+
+	tooltipIconLen = 32
```
Comment (ejulio-ts) [inline d2renderers/d2svg/d2svg.go:43]: `tooltipIconSize`?

File: main.go [hunk 2]
```diff
@@ -247,7 +248,8 @@ func compile(ctx context.Context, ms *xmain.State, plugin d2plugin.Plugin, sketc
 
 	out := svg
 	if filepath.Ext(outputPath) == ".png" {
-		svg := svg
+		svg := appendix.AppendTooltips(diagram, ruler, svg)
```
Comment (ejulio-ts) [inline main.go:251]: Why not call `d2svg` with an extra argument like `withTooltipAppendix` instead of manipulating raw HTML/SVG?

Comment (alixander) [reply main.go:251]: more composable. d2svg.Render would get real messy with all the flags.

File: d2renderers/d2svg/d2svg.go (hunk 1 above)
Comment (alixander) [reply d2renderers/d2svg/d2svg.go:43]: it's circular, and uses radius elsewhere, so length clarifies that

Comment (ejulio-ts) [reply d2renderers/d2svg/d2svg.go:43]: The radius usage is not related to this one.
maybe, diameter then

File: main.go (hunk 2 above)
Comment (ejulio-ts) [reply main.go:251]: I won't enter a long discussion on the composition as it seems quite hackish to split the html and append back, so I'll assume this is ok for now and we can refactor later on if we need something else

**Review by ejulio-ts — APPROVED** [review]

File: main.go (hunk 2 above)
Comment (alixander) [reply main.go:251]: yeah but that's two separate things. like it should not be an XML string until the very end. d2svg.Render should return a structured XML that postprocessing steps can manipulate.

Comment (ejulio-ts) [reply main.go:251]: it makes sense, though at the same time the interface could be actual composition/chain of functions where you don't need to parse, use regexes and change viewports after it was ready.
In this case, `main` would decide which functions to include in the pipeline
```
sag := Render([diagram, tooltips])

func diagram(buffer) {
    buffer.write(...)
}

func tooltips(buffer) {
    buffer.write(...)
}
```

Comment (alixander) [reply main.go:251]: > parse, use regexes and change viewports

it would still need to do all that in the proposed pseudocode

Comment (ejulio-ts) [reply main.go:251]: Why? if the buffer is always appending to the end.
`Render` opens `<svg>` and closes `</svg>` everything else happens in between

Comment (alixander) [reply main.go:251]: oh i see. yeah that'd be a good refactor to do.

File: d2renderers/d2svg/d2svg.go (hunk 1 above)
Comment (alixander) [reply d2renderers/d2svg/d2svg.go:43]: it is, it's replacing this icon

**Pull request merged**